### PR TITLE
feat: Shield protections added

### LINKED ISSUE #623: Add Support for AWS::Shield::*
- [ ] AWS::Shield::Protection
- [ ] AWS::ShieldRegional::Protection

### PR DESCRIPTION
**Pull request by amanenk**
Shield regional is skipped cause of lack of documentation.
e2e is skipped because of high cost of infrastructure.
closes #623

### CODE REVIEW

File: resources/services/shield/protections.go [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+			},
+			{
+				Name:        "tags",
+				Description: "The AWS tags of the resource.",
+				Type:        schema.TypeJSON,
+				Resolver:    ResolveShieldProtectionTags,
+			},
+			{
+				Name:        "application_layer_automatic_response_configuration_status",
+				Description: "Indicates whether automatic application layer DDoS mitigation is enabled for the protection.  This member is required.",
```
Comment (roneli) [inline resources/services/shield/protections.go:46]: remover member is required from descriptions.

Maybe shorten to application_layer_protection_status

Comment (roneli) [reply resources/services/shield/protections.go:46]: cc @bbernays

**Review by roneli — APPROVED** [review]
LGTM, added a few nits

**Review by roneli — CHANGES_REQUESTED** [review]
Would add the following APIs

https://docs.aws.amazon.com/waf/latest/DDOSAPIReference/API_DescribeSubscription.html)
https://docs.aws.amazon.com/waf/latest/DDOSAPIReference/API_ListProtectionGroups.html
https://docs.aws.amazon.com/waf/latest/DDOSAPIReference/API_ListAttacks.html <- would fetch only last 24 hours

**Comment by amanenk** [comment]
@roneli , @bbernays could you point me to `ShieldRegional` API? I can't find any documentation about it.

**Comment by bbernays** [comment]
> @roneli , @bbernays could you point me to `ShieldRegional` API? I can't find any documentation about it.

Added a note to the top issue, that we are skipping it as we don't have documentation of it in the SDK (or anywhere else)

File: resources/services/shield/protections.go [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+	})
+	if err != nil {
+		return diag.WrapError(err)
+	}
+
+	tags := map[string]*string{}
+	for _, t := range output.Tags {
+		tags[*t.Key] = t.Value
+	}
+	return resource.Set(c.Name, tags)
```
Comment (bbernays) [inline resources/services/shield/protections.go:121]: I think there is a `client` function to help with this functionality

**Review by bbernays — CHANGES_REQUESTED** [review]

File: resources/provider/migrations/postgres/28_v0.11.2.up.sql [hunk 3]
```diff
@@ -1,23 +1,23 @@
...
+    "description" text,
+    "source_location" text,
+    "source_repository" text,
+    "source_type" text,
+    "source_bundle_s3_bucket" text,
+    "source_bundle_s3_key" text,
+    "status" text,
+    "version_label" text,
+    CONSTRAINT aws_elasticbeanstalk_application_versions_pk PRIMARY KEY (arn),
+    UNIQUE (cq_id)
```
Comment (bbernays) [inline resources/provider/migrations/postgres/28_v0.11.2.up.sql]: Can you disable any auto linting by your IDE? Or maybe we can standardize across the team on the desired linting rules so we don't have changes to migrations outside of the initial commit that adds them

Comment (amanenk) [reply resources/provider/migrations/postgres/28_v0.11.2.up.sql]: I have managed to configure it as close as possible to the initial style. I could not find only `space before (` rule

File: resources/provider/provider.go [hunk 4]
```diff
@@ -3,6 +3,8 @@ package provider
 import (
 	"embed"
 
+	"github.com/cloudquery/cq-provider-aws/resources/services/shield"
+
```
Comment (roneli) [inline resources/provider/provider.go:7]: rogue  import line

File: resources/services/shield/protection_groups.go [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/service/shield"
+	"github.com/cloudquery/cq-provider-aws/client"
+	"github.com/cloudquery/cq-provider-sdk/provider/diag"
+	"github.com/cloudquery/cq-provider-sdk/provider/schema"
+)
+
+//go:generate cq-gen --resource protection_groups --config gen.hcl --output .
+func ProtectionGroups() *schema.Table {
```
Comment (roneli) [inline resources/services/shield/protection_groups.go:14]: Missing https://docs.aws.amazon.com/waf/latest/DDOSAPIReference/API_ListResourcesInProtectionGroup.html
 -> can add in separate PR (open issue)

File: resources/services/shield/protection_groups.go [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+				Type:        schema.TypeString,
+			},
+			{
+				Name:        "id",
+				Description: "The name of the protection group",
+				Type:        schema.TypeString,
+				Resolver:    schema.PathResolver("ProtectionGroupId"),
+			},
+			{
+				Name:        "arn",
```
Comment (roneli) [inline resources/services/shield/protection_groups.go:53]: ProtectionGroups doesn't have tags? (just verifying)

File: resources/services/shield/subscriptions.go [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+				Description: "If ENABLED, the Shield Response Team (SRT) will use email and phone to notify contacts about escalations to the SRT and to initiate proactive customer support",
+				Type:        schema.TypeString,
+			},
+			{
+				Name:        "start_time",
+				Description: "The start time of the subscription, in Unix time in seconds",
+				Type:        schema.TypeTimestamp,
+			},
+			{
+				Name:        "arn",
```
Comment (roneli) [inline resources/services/shield/subscriptions.go:76]: Same here do subscriptions have tags?

**Review by roneli — APPROVED** [review]
LGTM, had some comments, should open issue on missing API call, would verify the tag APIs on resources.

File: resources/services/shield/subscriptions.go (hunk 7 above)
Comment (amanenk) [reply resources/services/shield/subscriptions.go:76]: it seems that only protections and protection groups have tags
https://aws.amazon.com/about-aws/whats-new/2021/03/aws-shield-advanced-now-supports-resource-tagging/

File: resources/services/shield/protection_groups.go (hunk 5 above)
Comment (amanenk) [reply resources/services/shield/protection_groups.go:14]: protection_group has `members` column which tells which resources are in protection group

**Review by bbernays — APPROVED** [review]
LGTM

**Pull request merged**